### PR TITLE
Make GSGoal.QuestionClient work correctly at least for clients with ID < 2**16

### LINKED ISSUE #7534: GSGoal.QuestionClient sends message to wrong client
CmdGoalQuestion uses client index instead of client id but indexes are not synced on network so whoever receives it is somewhat random. 
Problem with sending client_id though is that it needs 32 bits and CmdGoalQuestion only have 27 available. 
Technically it can be squeezed there as only 10 bits is needed to send 988 possible button combinations but enumerating them isn't fun. 
So... any other ideas?

### PR DESCRIPTION
**Pull request by ldpl**
Partially fixes #7534. 
It's a good temporary fix as client ids almost never get that big and properly fixing it is much more complicated. 

Way to test:
1. start server with gs 
[testclientquestion.zip](https://github.com/OpenTTD/OpenTTD/files/3140298/testclientquestion.zip)
2. connect with another client (1)
3. connect with another client (2)
4. leave server with client (1)
5. connect with another client (3)
client 3 won't get welcome message without this fix (in 1.9.1)

### CODE REVIEW

**Comment by PeterN** [comment]
As `GOAL_QUESTION_BUTTON_COUNT` is 18, this leaves 14 bits available in p2, into which you could move the `type` and `is_client` parameters, and then use the full client ID.

**Comment by ldpl** [comment]
@PeterN  AFAICT full client_id is 32 bits > 27 = 13 + 14 so it won't be a full fix either. And I can't think of an easy way to squeeze buttons in 13 bytes even though it's theoretically possible (coz 1-3 bits only). I can send 16 bits of  client id though instead of 13 if you insist.

**Comment by nielsmh** [comment]
https://github.com/OpenTTD/OpenTTD/blob/a0c78c73e3c19dcbb1c295b30f17c38f7289b608/src/network/network_type.h#L17-L25
https://github.com/OpenTTD/OpenTTD/blob/a0c78c73e3c19dcbb1c295b30f17c38f7289b608/src/network/network_type.h#L40-L48
https://github.com/OpenTTD/OpenTTD/blob/a0c78c73e3c19dcbb1c295b30f17c38f7289b608/src/network/network_server.h#L21-L22

As far as I can tell none of this would allow a ClientID to actually break the 8 bit limit, despite the ClientID type likely mapping to `int` storage. Mainly since the pool defined in `network_server.h` uses 8 bit indices.

If you have a way to reproduce ClientIDs > 255 do share.

**Comment by michicc** [comment]
`ClientID` is not `ClientIndex`. The client ID is monotonically increased by the server for each connecting client and never reused.

https://github.com/OpenTTD/OpenTTD/blob/a0c78c73e3c19dcbb1c295b30f17c38f7289b608/src/network/network_server.cpp#L212-L212

File: src/goal.cpp [hunk 1]
```diff
@@ -236,40 +236,43 @@ CommandCost CmdSetGoalCompleted(TileIndex tile, DoCommandFlag flags, uint32 p1,
...
  * @return the cost of this operation or an error
  */
 CommandCost CmdGoalQuestion(TileIndex tile, DoCommandFlag flags, uint32 p1, uint32 p2, const char *text)
 {
 	uint16 uniqueid = (GoalType)GB(p1, 0, 16);
 	CompanyID company = (CompanyID)GB(p1, 16, 8);
-	ClientIndex client = (ClientIndex)GB(p1, 16, 8);
-	byte type = GB(p1, 24, 2);
-	bool is_client = HasBit(p1, 31);
+	ClientID client = (ClientID)GB(p1, 16, 16);
```
Comment (LordAro) [inline src/goal.cpp:251]: If `ClientIndex` is the "proper" type here (as discussed previously), why the change to `ClientID`? I'm not a fan of using arbitrary values for enum types, especially potentially well beyond their defined ranges, as certain aggressive gcc optimisations can truncate enum types or skip range checks...

Comment (michicc) [reply src/goal.cpp:251]: If I'm understanding the issue right, the problem is that `ClientIndex` is a local value and only `ClientID` is in fact synced across the network.

**Review by michicc — APPROVED** [review]
We did put it in 1.9.2, so...

**Pull request merged**